File: src/lib/events/feed-cache-ledger.ts
import { encodedJsonBytes } from '../cache/cache-byte-size';
import { cacheLedgerBytes } from '../cache/cache-ledger-bytes';
import { cacheLedgerId } from '../cache/cache-ledger-id';
import type { CacheLedgerRecord } from '../cache/cache-ledger-record';
import type { FeedScanHint } from './feed-scan-hints';
import type { FeedCoverage, FeedCursor } from './types';

export function feedCursorLedgerRecord(
  cursor: FeedCursor,
): CacheLedgerRecord {
  const draft = feedLedgerDraft('feed-page', 'feed-cursor', cursor);
  return withBytes(draft, cursor);
}

export function feedCoverageLedgerRecord(
  coverage: FeedCoverage,
): CacheLedgerRecord {
  const statusPenalty = coverage.status === 'complete' ? 500 : 100;
  const draft = feedLedgerDraft(
    'feed-coverage',
    'coverage-row',
    coverage,
    statusPenalty,
    coverage.relayUrl,
  );
  return withBytes(draft, coverage);
}

export function feedScanHintLedgerRecord(
  hint: FeedScanHint,
): CacheLedgerRecord {
  const draft = feedLedgerDraft(
    'feed-scan-hint',
    'scan-hint',
    {
      id: hint.id,
      feedKey: hint.scanKey,
      updatedAt: hint.updatedAt,
    },
    50,
    hint.relayUrl,
  );
  return withBytes(draft, hint);
}

function feedLedgerDraft(
  ownerKind: CacheLedgerRecord['ownerKind'],
  resourceKind: CacheLedgerRecord['resourceKind'],
  row: { readonly id: string; readonly feedKey: string; readonly updatedAt: number },
  baseScore = 300,
  relayUrl?: string,
): CacheLedgerRecord {
  return {
    id: cacheLedgerId(ownerKind, row.id),
    ownerKind,
    resourceKind,
    resourceId: row.id,
    score: baseScore + Math.floor(row.updatedAt / 3_600_000),
    createdAt: row.updatedAt,
    updatedAt: row.updatedAt,
    cacheBytes: 0,
    protected: false,
    feedKey: row.feedKey,
    relayUrl,
    reason: 'derived-feed-cache',
  };
}

function withBytes(
  draft: CacheLedgerRecord,
  row: unknown,
): CacheLedgerRecord {
  return {
    ...draft,
    cacheBytes: encodedJsonBytes(row) + cacheLedgerBytes(draft),
  };
}

File: crates/lkjstr-storage/src/feed_cache_ledger.rs
#![doc = "Feed cache ledger records for SQLite storage."]

use serde::Serialize;

use crate::{
    CacheLedgerRecord, FeedCoverageRecord, FeedCursorRecord, FeedScanHintRecord,
    resource::{CacheOwnerKind, CacheResourceKind},
    tab_state::{cache_ledger_id, encoded_json_bytes},
};

pub fn feed_cursor_ledger_record(
    row: &FeedCursorRecord,
) -> Result<CacheLedgerRecord, serde_json::Error> {
    ledger(row, CacheOwnerKind::FeedPage, CacheResourceKind::FeedCursor)
}

pub fn feed_coverage_ledger_record(
    row: &FeedCoverageRecord,
) -> Result<CacheLedgerRecord, serde_json::Error> {
    ledger(
        row,
        CacheOwnerKind::FeedCoverage,
        CacheResourceKind::CoverageRow,
    )
}

pub fn feed_scan_hint_ledger_record(
    row: &FeedScanHintRecord,
) -> Result<CacheLedgerRecord, serde_json::Error> {
    ledger(
        row,
        CacheOwnerKind::FeedScanHint,
        CacheResourceKind::ScanHint,
    )
}

fn ledger(
    row: &impl FeedLedgerRow,
    owner_kind: CacheOwnerKind,
    resource_kind: CacheResourceKind,
) -> Result<CacheLedgerRecord, serde_json::Error> {
    Ok(CacheLedgerRecord {
        id: cache_ledger_id(owner_kind, row.resource_id()),
        owner_kind,
        resource_kind,
        resource_id: row.resource_id().to_owned(),
        score: row.score(),
        created_at: row.updated_at_ms(),
        updated_at: row.updated_at_ms(),
        cache_bytes: encoded_json_bytes(row)?,
        protected: false,
        account_pubkey: None,
        feed_key: Some(row.feed_key().to_owned()),
        relay_url: row.relay_url().map(ToOwned::to_owned),
        reason: Some(row.reason().to_owned()),
    })
}

trait FeedLedgerRow: Serialize {
    fn resource_id(&self) -> &str;
    fn feed_key(&self) -> &str;
    fn updated_at_ms(&self) -> u64;
    fn score(&self) -> i64;
    fn reason(&self) -> &str;
    fn relay_url(&self) -> Option<&str>;
}

impl FeedLedgerRow for FeedCursorRecord {
    fn resource_id(&self) -> &str {
        &self.cursor_id
    }
    fn feed_key(&self) -> &str {
        &self.feed_key
    }
    fn updated_at_ms(&self) -> u64 {
        self.updated_at_ms
    }
    fn score(&self) -> i64 {
        20
    }
    fn reason(&self) -> &str {
        "feed-cursor"
    }
    fn relay_url(&self) -> Option<&str> {
        None
    }
}

impl FeedLedgerRow for FeedCoverageRecord {
    fn resource_id(&self) -> &str {
        &self.coverage_id
    }
    fn feed_key(&self) -> &str {
        &self.feed_key
    }
    fn updated_at_ms(&self) -> u64 {
        self.completed_at_ms
    }
    fn score(&self) -> i64 {
        if self.dense { 5 } else { 15 }
    }
    fn reason(&self) -> &str {
        "feed-coverage"
    }
    fn relay_url(&self) -> Option<&str> {
        Some(&self.relay_url)
    }
}

impl FeedLedgerRow for FeedScanHintRecord {
    fn resource_id(&self) -> &str {
        &self.hint_id
    }
    fn feed_key(&self) -> &str {
        &self.feed_key
    }
    fn updated_at_ms(&self) -> u64 {
        self.updated_at_ms
    }
    fn score(&self) -> i64 {
        10
    }
    fn reason(&self) -> &str {
        "feed-scan-hint"
    }
    fn relay_url(&self) -> Option<&str> {
        Some(&self.relay_url)
    }
}
